File: services/api/src/nova/schemas/memory.py
"""Memory payloads.

The embedding is deliberately absent from every schema here. It is an
implementation detail of retrieval, it is 1536 floats, and a client that
received it could do nothing with it.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from nova.models.memory import CATEGORIES

# Mirrors the model's check constraint. Declared as a Literal so an unknown
# category is a 422 with a useful message rather than a database error.
MemoryCategory = Literal["preference", "fact", "routine", "relationship", "project", "event"]

MAX_MEMORY_CONTENT = 300


class MemoryRead(BaseModel):
    """One thing NOVA remembers, as the owner sees it."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    content: str
    category: str
    importance: float
    confidence: float
    # Where it came from, so "why does it think that?" has an answer.
    source_conversation_id: uuid.UUID | None
    recall_count: int
    last_recalled_at: datetime | None
    created_at: datetime


class MemoryPage(BaseModel):
    """A page of memories with the total, so the app can show a count."""

    items: list[MemoryRead]
    total: int


class MemoryUpdate(BaseModel):
    """A correction from the person the memory is about.

    Every field is optional and unset means unchanged, so the app can send
    just what was edited.
    """

    model_config = ConfigDict(extra="forbid")

    content: str | None = Field(default=None, min_length=1, max_length=MAX_MEMORY_CONTENT)
    category: MemoryCategory | None = None
    importance: float | None = Field(default=None, ge=0.0, le=1.0)

    @field_validator("content")
    @classmethod
    def _normalise(cls, value: str | None) -> str | None:
        if value is None:
            return None
        collapsed = " ".join(value.split())
        if not collapsed:
            raise ValueError("Memory content must not be blank.")
        return collapsed


class MemorySearchResult(BaseModel):
    """A memory with how well it matched a query."""

    memory: MemoryRead
    # 1.0 is identical, 0.0 unrelated. Exposed because a retrieval screen
    # that shows scores is how someone judges whether NOVA is recalling the
    # right things.
    similarity: float


def _assert_categories_match() -> None:
    """Fail at import if the Literal has drifted from the model's tuple.

    The two have to be written out separately -- one is a type, the other a
    runtime tuple -- so this is what stops a category being added in one
    place and silently rejected by the other.
    """
    declared = set(MemoryCategory.__args__)  # type: ignore[attr-defined]
    if declared != set(CATEGORIES):  # pragma: no cover - unreachable while correct
        raise RuntimeError(
            f"MemoryCategory {sorted(declared)} does not match "
            f"nova.models.memory.CATEGORIES {sorted(CATEGORIES)}"
        )


_assert_categories_match()

File: apps/ios/NOVA/Models/Memory.swift
import Foundation

/// One thing NOVA remembers about its owner.
///
/// Deliberately readable rather than internal-looking: this is shown to the
/// person it is about, and the point of the screen is that they can tell
/// what NOVA believes and change it.
struct Memory: Codable, Identifiable, Equatable, Sendable {
    let id: UUID
    let content: String
    let category: MemoryCategory
    /// How much this shapes replies, 0-1.
    let importance: Double
    /// How sure NOVA is that it is true, 0-1. A correction sets it to 1.
    let confidence: Double
    /// Where it came from. Nil once that conversation has been deleted --
    /// the memory outlives the transcript.
    let sourceConversationId: UUID?
    let recallCount: Int
    let lastRecalledAt: Date?
    let createdAt: Date

    /// Never recalled reads differently from recalled once; the distinction
    /// is worth a word rather than a zero.
    var recallDescription: String {
        switch recallCount {
        case 0: "Never used"
        case 1: "Used once"
        default: "Used \(recallCount) times"
        }
    }
}

/// The closed taxonomy the API accepts.
///
/// `unknown` exists so a category added server-side does not stop the whole
/// list decoding -- a memory shown under a vague label is far better than a
/// screen that fails to load.
enum MemoryCategory: String, Codable, CaseIterable, Identifiable, Sendable {
    case preference, fact, routine, relationship, project, event
    case unknown

    var id: String { rawValue }

    /// Categories a person can choose. `unknown` is never offered.
    static var selectable: [MemoryCategory] {
        allCases.filter { $0 != .unknown }
    }

    init(from decoder: any Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = MemoryCategory(rawValue: raw) ?? .unknown
    }

    var label: String {
        switch self {
        case .preference: "Preference"
        case .fact: "Fact"
        case .routine: "Routine"
        case .relationship: "Person"
        case .project: "Project"
        case .event: "Event"
        case .unknown: "Other"
        }
    }

    var symbol: String {
        switch self {
        case .preference: "heart"
        case .fact: "text.book.closed"
        case .routine: "repeat"
        case .relationship: "person.2"
        case .project: "hammer"
        case .event: "calendar"
        case .unknown: "questionmark.circle"
        }
    }
}

/// A page of memories with the total, so a count can be shown without
/// fetching everything.
struct MemoryPage: Decodable, Sendable {
    let items: [Memory]
    let total: Int
}

/// A memory with how well it matched a search.
struct MemorySearchResult: Decodable, Identifiable, Sendable {
    let memory: Memory
    /// 1.0 is identical, 0.0 unrelated.
    let similarity: Double

    var id: UUID { memory.id }
}

// MARK: - Requests

/// A correction. Every field is optional: omitted means unchanged, so the app
/// sends only what was edited.
struct UpdateMemoryRequest: Encodable, Sendable {
    var content: String?
    var category: MemoryCategory?
    var importance: Double?
}
